internal/bootloader/grubenv/grubenv.go

```diff
@@ -0,0 +1,117 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package grubenv
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/canonical/x-go/strutil"
+)
+
+// FIXME: support for escaping (embedded \n in grubenv) missing
+type Env struct {
+	env      map[string]string
+	ordering []string
+
+	path string
+}
+
+func NewEnv(path string) *Env {
+	return &Env{
+		env:  make(map[string]string),
+		path: path,
+	}
+}
+
+func (g *Env) Get(name string) string {
+	return g.env[name]
+}
+
+func (g *Env) Set(key, value string) {
+	if !strutil.ListContains(g.ordering, key) {
+		g.ordering = append(g.ordering, key)
+	}
```

Comment:
To avoid an O(N) lookup here, can we use the map to check for existence instead?
```suggestion
	if _, present := g.env[key]; !present {
		g.ordering = append(g.ordering, key)
	}
```